Refactor official docs

MasterSkepticista opened the pull request:
#1219, with minor corrections.

### Vision: A cognitively easy to digest ground truth for everything OpenFL.

**Changelog:**

* Rename docs-requirements.txt to requirements.txt (already under docs folder)
* Update official docs contributing.md
* Point repo CONTRIBUTING.md to docs/contributing.md
* Add OpenFL logo and favicon
* **New:** Use new Sphinx Book Theme
* **New:** API reference now links to source code URL
* Remove modules.rst
* Add releases and contributing pages
* Removed rST prolog variables. Easier to use names directly.
* Roadmap is part of the documentation to avoid duplication.
* Follow NumPy NEP-29 (extends SPEC-0) python version support policy.
* User cognitive overhead reduction: For example, code block `$` signs removed. Code copy should only copy the snippet, not the `$` symbol.
* Remove requirement sorting in pre-commit checks (breaks readability otherwise)
* API toctree shown on left sidebar
* Fix sidebar toggle for using entire screen if necessary
* Recursively scan function and class docstrings for render (updated templates)
* **Preview:** Notebooks as part of documentation. This is a possible home for ALL `openfl-tutorials`.
* Up-to-date `installation` page.
* `sphinx-build` to be triggered directly in future. Command mentioned under `contributing.md`. `make` will work for now.
* **New:** Dark mode

Check live: https://openfl.readthedocs.io/en/karansh1-docs/index.html

MasterSkepticista committed:
Rebase from fork/feature to official/feature (#1222)

* Upgrade documentation (1/N)

* Rename docs-requirements.txt to requirements.txt (already under docs folder)
* Update official docs contributing.md
* Point repo CONTRIBUTING.md to docs/contributing.md
* Add OpenFL logo and favicon
* Use new Sphinx Book Theme
* [New] Code linker for API reference

Signed-off-by: Shah, Karan <karan.shah@intel.com>

* faster builds; update readme

Signed-off-by: Shah, Karan <karan.shah@intel.com>

* Restructure index

* Remove modules.rst
* Contribution guidelines moved to about/contributing
* Add releases and contributing pages

Signed-off-by: Shah, Karan <karan.shah@intel.com>

* Remove rST prolog variables. Easier to use names directly.

Signed-off-by: Shah, Karan <karan.shah@intel.com>

* Many more changes

* Update `contributing.md`
* Revamp `installation.md`
* Minor changes

Signed-off-by: Shah, Karan <karan.shah@intel.com>

* Misc

Signed-off-by: Shah, Karan <karan.shah@intel.com>

* Roadmap is part of docs

Signed-off-by: Shah, Karan <karan.shah@intel.com>

* Deprecation policy; recategorization

Signed-off-by: Shah, Karan <karan.shah@intel.com>

* Add taskrunner notebook auto-documenting example

Signed-off-by: Shah, Karan <karan.shah@intel.com>

* Update installation.md

Signed-off-by: Shah, Karan <karan.shah@intel.com>

* Add entire API docs, add update doc help in contributing.md

Signed-off-by: Shah, Karan <karan.shah@intel.com>

* Disable requirements sorting pre-commit check

Signed-off-by: Shah, Karan <karan.shah@intel.com>

* restore custom template for now

Signed-off-by: Shah, Karan <karan.shah@intel.com>

* API docs as toctree

Signed-off-by: Shah, Karan <karan.shah@intel.com>

* Fix sidebar toggle, delete out-of-date fonts and other files

Signed-off-by: Shah, Karan <karan.shah@intel.com>

* Restore templates; for real this time

Signed-off-by: Shah, Karan <karan.shah@intel.com>

* Update troubleshooting, migrate all diagrams under images/

Signed-off-by: Shah, Karan <karan.shah@intel.com>

* delete makefiles; use sphinx-build directly

Signed-off-by: Shah, Karan <karan.shah@intel.com>

* Multiple changes

* Batch rename of `console` -> `shell`
* Shrink titles
* Add `tutorials/` for notebooks
* Lot of reshuffling

Signed-off-by: Shah, Karan <karan.shah@intel.com>

* Increase nb exec timeout

Signed-off-by: Shah, Karan <karan.shah@intel.com>

* Prints recursive API docs finally!!!

Signed-off-by: Shah, Karan <karan.shah@intel.com>

* Restructure tutorial section

Signed-off-by: Shah, Karan <karan.shah@intel.com>

* Remove torch/torchvision from workflow req

Signed-off-by: Shah, Karan <karan.shah@intel.com>

* Revert make; exclude workflow notebook from execution

Signed-off-by: Shah, Karan <karan.shah@intel.com>

* Enhance taskrunner notebook

Signed-off-by: Shah, Karan <karan.shah@intel.com>

* Update index

Signed-off-by: Shah, Karan <karan.shah@intel.com>

* Updated notebooks

Signed-off-by: Shah, Karan <karan.shah@intel.com>

* Revert "Remove torch/torchvision from workflow req"

This reverts commit 5195e868f967129b9ca1a5f7ca8bdf8f33e8188c.

Signed-off-by: Shah, Karan <karan.shah@intel.com>

---------

Signed-off-by: Shah, Karan <karan.shah@intel.com>
Signed-off-by: MasterSkepticista <karan.shah@intel.com>

psfoley committed:
Remove fail on warning for readthedocs

Signed-off-by: Patrick Foley <psfoley@gmail.com>
Signed-off-by: MasterSkepticista <karan.shah@intel.com>

psfoley commented:
First of all, this is a vast improvement on our existing documentation @MasterSkepticista. Very well done!

I had to disable warnings on build, but here's a live version on ReadTheDocs:

https://openfl.readthedocs.io/en/karansh1-docs/index.html

A couple issues (I realize that not all of these were caused by this PR):
- The [Contributing](https://openfl.readthedocs.io/en/karansh1-docs/about/contributing.html) link is broken from the [Overview](https://openfl.readthedocs.io/en/karansh1-docs/index.html) page
- [Guide](https://openfl.readthedocs.io/en/latest/installation.html) link is broken on [Quickstart](https://openfl.readthedocs.io/en/karansh1-docs/tutorials/taskrunner.html):
- `image.png` link is broken on [MNIST Example: Workflow API](https://openfl.readthedocs.io/en/karansh1-docs/tutorials/workflow.html)
- On https://openfl.readthedocs.io/en/karansh1-docs/about/features_index/workflowinterface.html
    - [Features](https://openfl.readthedocs.io/en/karansh1-docs/about/features.rst) link is broken
- https://openfl.readthedocs.io/en/karansh1-docs/about/features_index/privacy_meter.html
- On the [Privacy Meter](https://openfl.readthedocs.io/en/karansh1-docs/about/features_index/privacy_meter.html) page 
    - [Here](https://github.com/securefederatedai/openfl/tree/f1657abe88632d542504d6d71ca961de9333913f/openfl-tutorials/experimental/workflow) link is broken

theakshaypant commented on docs/index.rst:
```diff
@@ -1,67 +1,107 @@
+.. note::
 
-Open Federated Learning (|productName|) is a Python\* \ 3 library for federated learning that enables organizations to collaboratively train a model without sharing sensitive information.
+   This project is continually being developed and improved. Expect changes to this manual, the project code, and the project design. We encourage community contributions! Refer to the `Contributing <about/contributing.html>`_ section for more information.
```
The contributing link is broken. Should be <contributing.html> as the relevant file is outside the `about` directory.

theakshaypant replied:
Resolved in fc630b0276848ff9817a11a377ccf041c9051c0b.

theakshaypant commented on docs/tutorials/taskrunner.ipynb:
This file is displayed when navigating to both, `Getting Started/Quickstart` and `Tutorials/Quickstart`. Ideally, we should keep only keep it once.

MasterSkepticista replied:
This specific one is by design. It nudges the user that both are same. 

Part of the intent with this new design is to bring out flaws in our APIs and documentation. Example: when you go to `Features > Workflow API`, it also gets highlighted under `Manual > Experimental Features > Workflow API`. This indicates us, developers, to remove redundant linking. 

This will be removed in follow up PRs (you too are welcome to add BTW!)

MasterSkepticista committed:
Fix all href, improve formatting

Signed-off-by: Shah, Karan <karan.shah@intel.com>
Signed-off-by: MasterSkepticista <karan.shah@intel.com>

theakshaypant commented on docs/about/features_index/taskrunner.rst:
```diff
@@ -515,17 +515,17 @@ Option 1: Deploy a Federation in a Docker Container
 
 .. note::
-    You have to built an |productName| image. See :ref:`install_docker` for details.
+    You have to built an OpenFL image. See :ref:`installation` for details.
```
^ You have to `build` an OpenFL image.

MasterSkepticista replied:
Rewrite of many docs is pending, contributions welcome :)

theakshaypant commented on docs/about/features_index/workflowinterface.rst:
```diff
@@ -122,7 +118,7 @@ Prior interfaces in OpenFL support the standard horizontal FL training workflow:
     5. The aggregator applies an aggregation function (weighted average, FedCurv, FedProx, etc.) to the model weights, and reports the aggregate metrics.
 
-The Task Assigner determines the list of collaborator tasks to be performed, and both in the task runner API as well as the interactive API these tasks can be modified (to varying degrees). For example, to perform federated evaluation of a model, only the aggregated_model_validation task would be selected for the assigner's block of the federated plan. Equivalently for the interactive API, this c ...
+The Task Assigner determines the list of collaborator tasks to be performed, and both in the task runner API as well as the interactive API these tasks can be modified (to varying degrees). For example, to perform federated evaluation of a model, only the aggregated_model_validation task would be selected for the assigner's block of the federated plan. Equivalently for the interactive API, this c ...
```
Last line, `Prior to OpenFL 1.5, there was not a great way to support this in OpenFL.`

theakshaypant commented on docs/developer_guide/running_the_federation_with_gandlf.rst:
Do we want to update the "Aggregator based workflow" to "TaskRunner API"?

MasterSkepticista replied:
Done https://github.com/securefederatedai/openfl/pull/1223/commits/a9319ec1b1885516d96bc6a2ef9cadec51a26051

theakshaypant commented on docs/developer_guide/structure/components.rst:
The figure at the end seems to be broken
```
.. figure:: ../../source/openfl/director_workflow.svg
```
<img width="849" alt="Screenshot 2024-12-19 at 11 41 33 AM" src="https://github.com/user-attachments/assets/a63d4219-c57c-466c-87e2-919283b5e39f" />

MasterSkepticista replied:
Fixed in https://github.com/securefederatedai/openfl/pull/1223/commits/25c6e34cc5244ed077297baef24ccb3bdaaffe91

theakshaypant commented on docs/about/features_index/privacy_meter.rst:
The example link at the bottom of the page is broken.

MasterSkepticista replied:
Fixed in https://github.com/securefederatedai/openfl/commit/cf844bc7127a78c118ab023f3fcc7f8dca1e3c92

theakshaypant reviewed (commented):
The new styling looks amazing. Love the light/dark theme choice. 

Have a few comments, some of them might not be related to the changes made in this PR but issues with the documentation as I was going through the readthedocs hosting alongside it.

theakshaypant commented:
Hijacking this PR to also note that the [Director-Envoy Communication](https://openfl.readthedocs.io/en/karansh1-docs/developer_guide/structure/communication.html#director-envoy-communication) mermaid diagram isn't very clear in dark mode.

MasterSkepticista committed:
Fix diagram url

Signed-off-by: MasterSkepticista <karan.shah@intel.com>

MasterSkepticista committed:
Rename to TaskRunner

Signed-off-by: MasterSkepticista <karan.shah@intel.com>

MasterSkepticista commented:
Thanks for the feedback @theakshaypant. Addressed relevant items.
Remainder, will be part of future PRs. Some of it is simple deletion of deprecated API docs, and remainder needs rewrite + restructuring, as the scope here was to get the ball rolling.

> First of all, this is a vast improvement on our existing documentation @MasterSkepticista. Very well done!
> 
> I had to disable warnings on build, but here's a live version on ReadTheDocs:
> 
> https://openfl.readthedocs.io/en/karansh1-docs/index.html
> 
> A couple issues (I realize that not all of these were caused by this PR):
> 
> * The [Contributing](https://openfl.readthedocs.io/en/karansh1-docs/about/contributing.html) link is broken from the [Overview](https://openfl.readthedocs.io/en/karansh1-docs/index.html) page
> * [Guide](https://openfl.readthedocs.io/en/latest/installation.html) link is broken on [Quickstart](https://openfl.readthedocs.io/en/karansh1-docs/tutorials/taskrunner.html):
> * `image.png` link is broken on [MNIST Example: Workflow API](https://openfl.readthedocs.io/en/karansh1-docs/tutorials/workflow.html)
> * On https://openfl.readthedocs.io/en/karansh1-docs/about/features_index/workflowinterface.html
>   
>   * [Features](https://openfl.readthedocs.io/en/karansh1-docs/about/features.rst) link is broken
> * https://openfl.readthedocs.io/en/karansh1-docs/about/features_index/privacy_meter.html
> * On the [Privacy Meter](https://openfl.readthedocs.io/en/karansh1-docs/about/features_index/privacy_meter.html) page
>   
>   * [Here](https://github.com/securefederatedai/openfl/tree/f1657abe88632d542504d6d71ca961de9333913f/openfl-tutorials/experimental/workflow) link is broken

Addressed most. The only ones I cannot test right now, are notebook URLs. They cannot be relative to the `docs`, only complete URLs, AFAIU. I have updated the links to be what they "will be" post-merge. We will do hotfixes if necessary.

theakshaypant reviewed (approved):
@MasterSkepticista Agreed on your response that rewrite/restructure should be covered separately.
This LGTM having gone through it a couple of times.

tanwarsh commented on docs/about/features.rst:
```diff
@@ -1,7 +1,3 @@
-.. # Copyright (C) 2020-2024 Intel Corporation
```
NIT: In some places, we are removing this header, while in others, it is still present. Can we standardize it one way or the other?

MasterSkepticista replied:
I removed headers in files I touched; no reason why not to remove for all docs...
At times it is either out of date, or not in the format as consistent across the framework. An elegant approach is to have a pre-commit check that applies the correct template on all files (and warns if missing or incorrect). 

Could you check if we can integrate a pre-commit tool that updates all headers?

theakshaypant replied:
Could perhaps check pre-commit hooks such as [this](https://github.com/Quantco/pre-commit-mirrors-insert-license) one.

MasterSkepticista replied:
This may work for updating headers, but won't check (?).
I came across this [go based tool](https://github.com/google/addlicense). Will it be easy to integrate in CI, or as pre-commit check?

tanwarsh reviewed (approved):
Great work @MasterSkepticista. LGTM. Please feel free to dismiss the comment if you think it is not necessary.

teoparvanov commented on docs/index.rst:
```diff
@@ -1,67 +1,105 @@
-   get_started/quickstart
-   get_started/examples
+   tutorials/taskrunner
+   tutorials/workflow
```
I suggest removing the "Tutorials" section, and moving all three intro items under "Getting started":
```
installation
tutorials/taskrunner
tutorials/workflow
```

MasterSkepticista replied:
I'm thinking of `Tutorials` as the staging area for all notebooks housed in OpenFL and advanced user guides :)

A feature of this new documentation upgrade is that notebooks are auto-executed during builds. The output is captured and part of the rendered HTML. It is also an implicit CI. We can revisit breakdown of sections as more content lines up in future PRs.

teoparvanov commented on docs/tutorials/taskrunner.ipynb:
```diff
@@ -0,0 +1,266 @@
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# Quickstart\n",
```
How about `Quickstart with TaskRunner API` ?

MasterSkepticista replied:
I think this exposes jargon a little too early. A user at that stage has _just_ installed OpenFL. So I'd prefer keeping the name. WDYT?

That said, the guide itself falls short of ideal expectations IMO - We are discussing CA and TLS steps in the guide which are not really "quickstart", rather an "advanced guide". But that is for a separate PR.

teoparvanov commented on docs/tutorials/workflow.ipynb:
```diff
@@ -0,0 +1,1798 @@
+   "id": "14821d97",
+   "metadata": {},
+   "source": [
+    "# MNIST Example: Workflow API"
```
nit - `An MNIST example with Workflow API` ?

teoparvanov reviewed (approved):
These are great improvements, thank you @MasterSkepticista ! My only suggestion is to consider removing the duplication of the link to the Quickstart in the index (left menu). By the way, the Quickstart page takes quite a while to load (could it be because it's present twice in the index)?

teoparvanov commented on docs/index.rst:
```diff
@@ -1,67 +1,105 @@
-   get_started/quickstart
-   get_started/examples
+   tutorials/taskrunner
+   tutorials/workflow
```
By the way, there are some links to those on the main [README.md's Getting Started](https://github.com/securefederatedai/openfl/blob/develop/README.md#getting-started) section. 

I guess those would need to be updated in a subsequent PR, after publishing the latest docs to readthedocs?

MasterSkepticista replied:
Good catch. I will include relevant changes in this PR.
Edit: I think those links will continue to work.

MasterSkepticista commented:
> These are great improvements, thank you @MasterSkepticista ! My only suggestion is to consider removing the duplication of the link to the Quickstart in the index (left menu). By the way, the Quickstart page takes quite a while to load (could it be because it's present twice in the index)?

@theakshaypant had a similar https://github.com/securefederatedai/openfl/pull/1223#discussion_r1891039557. Since this concern has appeared twice, I have made relevant changes in https://github.com/securefederatedai/openfl/pull/1223/commits/b73d487afa8d207daa4bf4077da72cbdfeb29358

@teoparvanov on your second question: I don't see any lag on loading the quickstart. This may be due to this being a dev build, or network issues.

MasterSkepticista committed:
Remove redundant quickstart under tutorials

Signed-off-by: MasterSkepticista <karan.shah@intel.com>

kminhta reviewed (approved):
LGTM! I took a few passes and couldn't find any additional broken links or egregious typos

Fantastic contribution @MasterSkepticista ! This is an excellent foundation for future doc improvements

> @teoparvanov on your second question: I don't see any lag on loading the quickstart. This may be due to this being a dev build, or network issues.

Also, no latency issue loading the quickstart on my end either
